### ISSUE TITLE
How to do directional transform?

### ISSUE THREAD

**Issue by cloudhan** [egraphs-good/egglog#256]
Say I have a `a*b + c` and I then decide to add an `fma` instruction to my program. I want to directionally **fuse** the original expr as `fma(a, b c)` because it is faster.

Maybe assigning them with properly crafted costs will solve the problem. Then this becomes an unmanageable ordering problem.

In a real world program, especially in the field of deep learning, there are numerious fusion patterns can be used. Properly ordering all of then with cost is not pratical. So the idea here is to decompose all pattern into steps, apply them one by one (with human insight on the order of applying).

```
--->: apply fusion then extract or discard (don't change the program) if not apply

original program --> program with fusion pattern 1 --> program with fusion pattern 1 and 2 --> ...
````

And this also allows us to maintain the rules of each step seperately in a standalone library/file.

Is this kind of directional transfrom achievable? If yes how?

[1] [ægraphs: Acyclic E-graphs for Efficient Optimization in a Production Compiler](https://pldi23.sigplan.org/details/egraphs-2023-papers/2/-graphs-Acyclic-E-graphs-for-Efficient-Optimization-in-a-Production-Compiler)
[2] https://github.com/bytecodealliance/rfcs/blob/main/accepted/cranelift-egraph.md

**Comment by saulshanabrook** [comment]
So am I understanding the example correctly that you have some set of fused operators, and it isn't practical to assign each a set cost, but you could order them in some way? And when you apply them in an order, you always want to extract out the version with them applied?

I have had some luck with the Python bindings basically making a set of rules, applying them, extracting out the best candidate, then creating a new e-graph with those same rules (but with all their costs much higher) and some new set of rules and constructs, applying them, and extracting out again. I am not sure if this is what you are trying to do.

If you had an example of where a simple cost function wouldn't work, but you have certain behavior in mind, I could try to see if there is a way to do it.

**Comment by cloudhan** [comment]
Yes, that is exactly what I am expecting. Do you have some minimal example to show how each parts should be organized?

**Comment by saulshanabrook** [comment]
I do something like this in the Array API example in the Python bindings.

I have to rewrite the `mean` and `std` functions as definitions, because Numba doesn't support those higher level functions when the `axis` term is provided ([link to rules](https://github.com/metadsl/egglog-python/blob/e0065e42a7d527e3d4bfccad0810236c14718b6a/python/egglog/exp/array_api_numba.py#L21-L38)).

The way I do it currently is by making the higher level functions I want to rewrite into low level ones have a much higher cost ([link to function def](https://github.com/metadsl/egglog-python/blob/e0065e42a7d527e3d4bfccad0810236c14718b6a/python/egglog/exp/array_api.py#L1226-L1228)).

This way, even though more ops are added, those will be extracted.


I end up applying those by creating a new e-graph based on all the existing rules + the specific rules for the numba translations, see [this example in the docs](https://egg-smol-python.readthedocs.io/en/stable/tutorials/sklearn.html#translating-for-numba).

Happy to answer other questions on the example, it's a bit in nuanced.

---

It's unfortunate though in that example that I have to change the original function cost, even though the optimization is in a different module.  Ideally the separate module with the optimization/rewrite should be able to define that it's "directional" as you put it, so that we only use the new version, instead of the old one.

I have a proposal to make this more straightforward, @oflatt might have been the one with this idea, although I can't find our conversation. There is currently an `unextractable` parameter you can use to mark a certain function as never able to be extracted. I would propose extending that as well to individual function calls, so that every function call could be additionally marked as unextractable. This could be done through an action, to be used in `rule`s, as well as a parameter to `rewrite`, which would mark the source as unextractable after matching.

Someone also just asked how I would do a rewrite for predicate pushdowns, i.e. `exp(vec)[idx] -> exp(vec[idx])`. Of course, you can do this today, but to make it consistently extract out the pushed-down term would require some more advanced cost model. Or, with this proposal, the source expression would be marked unextractable after its been matched.

**Comment by oflatt** [comment]
I think some sort of "don't extract this function call (enode)" would be useful. More generally, we could give a cost to individual function calls- this is also related to some ideas about subsumption we've had.

However, we may not even need to bake this into the back-end of egglog. Extraction can be done using egglog rules, building terms and putting them in a table. (see [this pr](https://github.com/oflatt/eggcc/pull/154) I'm working on)

Then your extraction rules could take into account your costs.

**Comment by saulshanabrook** [comment]
Ah yeah I see in your PR where you implement [a lowest cost node/cost](https://github.com/oflatt/eggcc/pull/154/files#diff-df9699d107787c7ca29cd610c9bbf2f9ce9770f7466c4abf8b9cda4ef310d234R7-R34) function, by using a custom merge function.

I had thought of doing something similar, but was worried about using custom merge functions for e-nodes, I wasn't sure if that would blow up or not in the way you are doing it. Cool that you got it working!

I might give implementing that "don't extract this function call (enode)" feature a go this week, so there is a simple way of implementing extracting [these sorts of rewrites](https://egg-smol-python.readthedocs.io/en/latest/explanation/indexing_pushdown.html).
